ws: Move to /etc/ssh/ssh_known_hosts

martinpitt opened the pull request:
There is no real reason for maintaining our own
/var/lib/cockpit/known_hosts file, as ssh itself already has a global
one in /etc/ssh/ssh_known_hosts. Use that by default, but fallback to
the legacy file for (1) lookups if a host is not already known in the
former but known in the latter; and (2) for writing if the ws we talk to
is still an old version (by checking if ws still has the "ssh"
capability).

Move the determination and setting of the known hosts file into a new
set_knownhosts_file() function, as it is now reasonably complex, will
be extended further in the future with more sources of known hosts, and
avoids handling SSH_OPTIONS_KNOWNHOSTS in multiple different places.

Adjust the integration tests to the new path and add new tests for
covering the fallback to the legacy file.

----

Blockers:
 - [x] land PR #6074 
 - [ ] land Peter's cockpit-ssh splitout
 - [ ] adjust commit message as appropriate (for example "ws:" prefix might not be appropriate any  more)

petervo commented on src/ws/ssh.c:
```diff
@@ -1207,23 +1208,32 @@ cockpit_ssh_connect (CockpitSshData *data,
+              data->logname,
+              data->ssh_options->knownhosts_file, PACKAGE_LOCALSTATE_DIR "/known_hosts");
+      data->ssh_options->knownhosts_file = PACKAGE_LOCALSTATE_DIR "/known_hosts";
+      host_known = cockpit_is_host_known (data->ssh_options->knownhosts_file, host, port);
```
I would only change the known host file option if the machine is found in /var/lib/cockpit/known_hosts. If it isn't found we should leave the setting untouched. Also need to better way to detect if we are using tests as it's valid for non test use cases to change the knownhost file using env vars.

martinpitt replied:
Hmm, i think it might actually be preferrable to compare `data->ssh_options->knownhosts` to the complete default path, not just `/etc`. I. e. as soon as the caller specifies something custom we don't want the fallback.

petervo replied:
Yeah I guess that makes sense, but we still should leave the setting untouched and only change it if a match is found in the fallback.

martinpitt replied:
Agreed, that's a good idea.

petervo commented on src/ws/ssh.c:
```diff
@@ -1207,23 +1208,32 @@ cockpit_ssh_connect (CockpitSshData *data,
+    {
+      g_debug ("%s: %s is not known in %s, falling back to %s",
+              data->logname,
+              data->ssh_options->knownhosts_file, PACKAGE_LOCALSTATE_DIR "/known_hosts");
```
maybe make this a const at the top, legacy_known_hosts or something

martinpitt commented:
Adjusted to Peter's feedback above. I also moved the code around a bit so that the entire logic of looking at various "known hosts" files and setting the right one is now in just one function `set_knownhosts_file()`. We can add `~/.ssh/known_hosts` there or write a temporary file for grabbing a host key from `/etc/cockpit/machines.d/*.json`. I didn't do that yet (one step at a time), but it's now straightforward to do.

Unlike for `machines.json` there is no hurry for landing the migration of the existing `/var/lib/cockpit/known_hosts` in lockstep, as we always have to assume and get along with an already existing file in `/etc`, and also can't assume that we are able to write it (as we don't "own" it), so I propose to do the migration in a separate PR. I was wondering where to actually do that -- as far as I can tell there are these options:

 1. through a root command  from JavaScript; I would add this to pkg/lib/machines.js `add_key()`, as there we change `ssh_known_hosts` anyway.
 2. to the (privileged) bridge, doing this migration along with `machines.json`
 3. don't bother and look at the old path forever

martinpitt commented:
Most test failures are unrelated, but `testBasic (check_multi_machine_key.TestMultiMachineKeyAuth)` on the *-atomics looks relevant:
```
not ok 103 testBasic (check_multi_machine_key.TestMultiMachineKeyAuth) duration: 138s
Traceback (most recent call last):
  File "./verify/check-multi-machine-key", line 158, in testBasic
    b.wait_in_text('#troubleshoot-dialog h4', "Log in to")
  File "/build/cockpit/test/common/testlib.py", line 243, in wait_in_text
    return self.wait_js_func('ph_in_text', selector, text)
  File "/build/cockpit/test/common/testlib.py", line 210, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 729, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 755, in _invoke
    raise Error(res['error'])
Error: timeout
Untrusted host
```

Corresponding journal messages:
```
Mar 06 08:15:05 localhost.localdomain.localdomain dockerd-current[1449]: MESSAGE: cockpit-protocol: failed to open known hosts file /etc/ssh/ssh_known_hosts
Mar 06 08:15:05 localhost.localdomain.localdomain dockerd-current[1449]: MESSAGE: cockpit-protocol: failed to open known hosts file /var/lib/cockpit/known_hosts
Mar 06 08:15:05 localhost.localdomain.localdomain dockerd-current[1449]: MESSAGE: Unknown: cockpit-ssh admin@10.111.115.185:22: ssh-ed25519 host key for server is not known: eb:b1:d2:22:12:f7:ad:3f:6f:16:8b:2d:07:a8:3c:6b
Mar 06 08:15:05 localhost.localdomain.localdomain dockerd-current[1449]: MESSAGE: cockpit-protocol: failed to open known hosts file /etc/ssh/ssh_known_hosts
Mar 06 08:15:05 localhost.localdomain.localdomain dockerd-current[1449]: MESSAGE: cockpit-protocol: failed to open known hosts file /var/lib/cockpit/known_hosts
Mar 06 08:15:05 localhost.localdomain.localdomain dockerd-current[1449]: MESSAGE: Unknown: cockpit-ssh admin@10.111.115.185:22: ssh-ed25519 host key for server is not known: eb:b1:d2:22:12:f7:ad:3f:6f:16:8b:2d:07:a8:3c:6b
```

This might need an SELinux policy change (@petervo alluded to that during our discussion), looking into that.

martinpitt commented:
Some more info about the Atomic failure: on the host `/etc/ssh/ssh_known_hosts` gets written properly, but this is not available to cockpit-ssh that is running in the cockpit ws container. Without this PR, `/var/lib/cockpit/known_hosts` exists in the container as well, as `/var/ is bind-mounted into the container. So `atomic-run` needs to be changed to bind-mount /etc/ssh/ssh_known_hosts too (or just `/etc/ssh/` wholesale, as there's no ssh in the container anyway).

martinpitt commented:
I pushed another commit on top which supposedly updates the container build scripts. However, I still haven't found out how to test this locally, as the container is not actually (re)built in `testsuite-prepare`, that only [updates the RPMs](https://github.com/cockpit-project/cockpit/blob/master/test/images/scripts/lib/atomic.install#L171). So this will very likely fail still.

Issues:
 * How to update the ws container image for that PR? (Will ask in meeting)
 * Do we need to be concerned about running an old cockpit/ws container on an Atomic host with cockpit ≥ 134, i. e. where `machines.json` writes to the *host* file system `/etc/ssh/ssh_known_host`, but the older ws container doesn't see this? The dashboard isn't actually visible on Atomic, so this might be a practical non-issue, but our `test/verify/check-multi-machine-key` test case exercises this.

martinpitt commented:
> Do we need to be concerned about running an old cockpit/ws container on an Atomic host with cockpit ≥ 134

Apparently yes. Meeting decision was to 
[introduce and check for a new ws capability](https://github.com/cockpit-project/cockpit/blob/master/src/ws/cockpitwebservice.c#L1644) and write the old file if that's absent.

> How to update the ws container image for that PR?

This is obsolete then. CI would cover the "legacy fallback" path on Atomic until we release Cockpit 134 and update the container, and from then on test writing the file to `/etc`.

martinpitt commented:
> > How to update the ws container image for that PR?
> This is obsolete then.

Argh, actually it's not. `testsuite-prepare` (through `atomic.install`) updates the  RPMs in the test image, but nothing else. So as  a short-term fix for that CI problem I propose to update `atomic-run` along with the RPMs.

In the long term we should make up our minds though if CI is supposed to test the latest bits (then we should update/rebuild the entire container) or backwards compatibility with the last public container image (then we shouldn't touch it at all), or preferrably even both somehow.

martinpitt commented:
I added another commit to update the cockpit/ws `/container/atomic-*` scripts along with updating the RPMs, so that we don't  run out of sync between those. This is less intrusive than completely reworking the container rebuild (this should happen in a separate PR, and isn't time critical, unlike this PR). Atomic tests now succeed locally for me, so I think this is ready for re-review.

martinpitt commented:
For the record: Our CI cannot test the fallback to writing `/var/lib/cockpit/known_hosts` if running against an older ws. I tested this manually and it works as expected.

petervo commented:
> For the record: Our CI cannot test the fallback to writing
> |/var/lib/cockpit/known_hosts| if running against an older ws. I tested
> this manually and it works as expected.
>

Can't we do this manually in either a unit or integration test?

martinpitt commented:
> Can't we do this manually in either a unit or integration test?

In an unit test this is impractical [1], but with integration tests it would work in principle: we could run all atomic tests both against the `cockpit/ws` container (which we updated to new RPMs and scripts) and once more against the `docker.io/cockpit/ws` container that got shipped on the original image. That might be a little too much drive-by work for this PR, but I can look into adding an explicit new `checkLastReleasedContainer()` test case to `check-multi-machine` which is only run on the Atomic OSes and modify testvm.py's `start_cockpit()` to be able to start the unmodified container.

[1] unless we create/use a mock c-ws which does not export the new capability, and a mock c-ssh which only reads the file in `/var` - but this would then not prove much, and is a lot of overhead.

petervo commented:
Ah i see, let's hold off on any changes and see where we land with the capabilities.

martinpitt commented:
This is scheduled for the current sprint, and important for long-term support (on-disk format change), thus marking as priority so that we make sure this (or a variation) lands in the next release.

petervo commented:
Ok, since we will be dropping the ssh capability from cockpit-ws all together. I think it's best to drop the new capability and do something like.

```
-        mod.allow_connection_string = $.inArray("connection-string", caps) != -1;
-        mod.has_auth_results = $.inArray("auth-method-results", caps) != -1;
+        /* If cockpit-ws is handling ssh, check for each capability. Otherwise
+         * the version is new enough that is has them all */
+        if ($.inArray("ssh", caps) > -1) {
+            mod.allow_connection_string = $.inArray("connection-string", caps) != -1;
+            mod.has_auth_results = $.inArray("auth-method-results", caps) != -1;
+            mod.known_hosts_path = "/var/lib/cockpit/known_hosts";
+        } else {
+            mod.allow_connection_string = true;
+            mod.has_auth_results = true;
+        }
```

Of course that means the PR will be blocked on the ssh move, but i think that's ok.

martinpitt commented:
Adjusted to Peter's suggestion above, nice one! I updated the description to enumerate the blockers, and label accordingly - right now CI would fail as the new test would expect `/etc`, but due to checking for the absence of the `ssh` cap it would write into `/var`. I'll verify that everything works as it should after Peter lands his work.

martinpitt commented:
This is now part of PR #6102. It depends on that rework to avoid having to introduce a temporary ws capability, but it should also not land after that PR as it would then require a new capability.
